core/src/main/scala/org/apache/spark/scheduler/Stage.scala

```diff
@@ -92,6 +92,35 @@ private[spark] abstract class Stage(
    */
   private var _latestInfo: StageInfo = StageInfo.fromStage(this, nextAttemptId)
 
+  /**
+   * Spark is resilient to executors dying by retrying stages on FetchFailures. Here, we keep track
+   * of unique stage failures (per stage attempt) triggered by fetch failures to prevent endless
+   * stage retries. Specifically, per stage we wish to only record a failure when the following
+   * holds:
+   *
+   * A) A fetch failure was observed
+   * B) A failure has not yet been registered for this stage attempt. There may be multiple
+   * concurrent failures for a sinlge stage since we may have multiple tasks executing at the same
+   * time, one or many of which may fail. Also, even though there may only be one non-zombie stage
+   * attempt, zombie stage attempts may still have running tasks.
+   */
```

Comment:
This comment is super verbose, and after reading it, it's still not obvious what this is actually storing.  What about something like:

Set of IDs of stage attempts that have failed with a FetchFailure.  We keep track of these failures in order to avoid endless retries if a stage keeps failing with a FetchFailure.  We keep track of each attempt ID that has failed to avoid recording duplicate failures if multiple tasks from the same stage attempt fail.